lib/src/main/java/tech/relaycorp/relaydroid/ThirdPartyEndpoint.kt

```diff
@@ -1,5 +1,109 @@
 package tech.relaycorp.relaydroid
 
-public sealed class ThirdPartyEndpoint(override val address: String) : Endpoint
-public class PrivateThirdPartyEndpoint(address: String) : ThirdPartyEndpoint(address)
-public class PublicThirdPartyEndpoint(address: String) : ThirdPartyEndpoint(address)
+import org.bouncycastle.asn1.x500.style.BCStyle
+import tech.relaycorp.relaydroid.storage.persistence.PersistenceException
+import tech.relaycorp.relaynet.RelaynetException
+import tech.relaycorp.relaynet.issueDeliveryAuthorization
+import tech.relaycorp.relaynet.wrappers.x509.Certificate
+import tech.relaycorp.relaynet.wrappers.x509.CertificateException
+import java.security.PublicKey
+import java.time.ZonedDateTime
+
+public sealed class ThirdPartyEndpoint(
+    override val thirdPartyAddress: String
+) : Endpoint {
+
+    public companion object {
+
+        // Private
+
+        @Throws(PersistenceException::class)
+        public suspend fun loadPrivate(
+            firstPartyAddress: String, thirdPartyAddress: String
+        ): PrivateThirdPartyEndpoint? {
+            val key = "${firstPartyAddress}_$thirdPartyAddress"
+            return Storage.privateThirdPartyAuthorization.get(key)?.let {
+                PrivateThirdPartyEndpoint(firstPartyAddress, thirdPartyAddress, it)
+            }
+        }
+
+        @Throws(
+            PersistenceException::class,
+            InvalidFirstPartyEndpointAddressException::class
+        )
+        public suspend fun importPrivateAuthorization(pda: Certificate): PrivateThirdPartyEndpoint {
```

Comment:
I missed two very important things in my review earlier:

- A `PrivateThirdPartyEndpoint` wouldn't be complete without the identity certificate of that third-party endpoint. We need it to encrypt parcels.
- We must validate that that the `pda` was actually issued by that identity certificate. If we blindly take the CommonName of the issuer, an attacker could take advantage of that to break the channel between two endpoints.

To fix the latter, we need to do something like this:

```kotlin
try {
    pda.getCertificationPath(emptyList(), listOf(identityCertificate))
} catch (exc: CertificateException) {
    throw InvalidAuthorizationException("PDA was not issued by third-party endpoint", exc)
}
```

That's why I didn't originally implement `Certificate.issuerCommonName`, but I didn't remember. So I'm going to remove `issuerCommonName`.